src-java/connecteddevices-topology/connecteddevices-storm-topology/src/main/java/org/openkilda/wfm/topology/connecteddevices/service/PacketService.java

```diff
@@ -75,52 +80,71 @@ public PacketService(PersistenceManager persistenceManager) {
+        SwitchConnectedDevice device;
 
-            if (device == null) {
-                return;
-            }
+        if (cachedDevice != null) {
+            device = cachedDevice;
+        } else {
+            Optional<SwitchConnectedDevice> optionalDevice = switchConnectedDeviceRepository
+                    .findLldpByUniqueIndex(createUniqueLldpIndex(data, flowRelatedData.getOriginalVlan()));
+            device = optionalDevice.orElse(null);
+        }
 
-            device.setTtl(data.getTtl());
-            device.setPortDescription(data.getPortDescription());
-            device.setSystemName(data.getSystemName());
-            device.setSystemDescription(data.getSystemDescription());
-            device.setSystemCapabilities(data.getSystemCapabilities());
-            device.setManagementAddress(data.getManagementAddress());
-            device.setTimeLastSeen(Instant.ofEpochMilli(data.getTimestamp()));
-            device.setFlowId(flowRelatedData.flowId);
-            device.setSource(flowRelatedData.source);
-        });
+        if (device != null) {
+            transactionManager.doInTransaction(() -> {
+                device.setTtl(data.getTtl());
+                device.setPortDescription(data.getPortDescription());
+                device.setSystemName(data.getSystemName());
+                device.setSystemDescription(data.getSystemDescription());
+                device.setSystemCapabilities(data.getSystemCapabilities());
+                device.setManagementAddress(data.getManagementAddress());
+                device.setTimeLastSeen(Instant.ofEpochMilli(data.getTimestamp()));
+                device.setFlowId(flowRelatedData.getFlowId());
+                device.setSource(flowRelatedData.getSource());
+            });
+        } else {
+            createLldpDevice(data, flowRelatedData);
+        }
     }
 
     /**
      * Handle Arp info data.
      */
     public void handleArpData(ArpInfoData data) {
-        transactionManager.doInTransaction(() -> {
-
-            FlowRelatedData flowRelatedData = findFlowRelatedData(data);
-            if (flowRelatedData == null) {
-                return;
-            }
+        FlowRelatedData flowRelatedData = findFlowRelatedData(data);
+        if (flowRelatedData == null) {
+            return;
+        }
 
-            SwitchConnectedDevice device = getOrCreateArpDevice(data, flowRelatedData.originalVlan);
+        SwitchConnectedDevice cachedDevice = switchConnectedDeviceCache.get(createUniqueArpIndex(data,
+                flowRelatedData.originalVlan));
+        SwitchConnectedDevice device;
 
-            if (device == null) {
-                return;
-            }
+        if (cachedDevice != null) {
+            device = cachedDevice;
+        } else {
+            Optional<SwitchConnectedDevice> optionalDevice = switchConnectedDeviceRepository
+                    .findArpByUniqueIndex(createUniqueArpIndex(data, flowRelatedData.originalVlan));
```

Comment:
minor: I think you can save createUniqueArpIndex into some variable just to do not call it two times